ops/_private/json.py

```diff
@@ -0,0 +1,38 @@
+# Copyright 2021 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# autopep8 hates monkeypatching
+# autopep8: off
+# flake8: noqa
+
+"""Internal JSON helpers."""
+
+from json import JSONEncoder
+
+# Explicitly import to avoid circular ImportErrors
+import ops.framework
+
+# Monkeypatch so Stored[Dict|List|Set] can be serialized
+
+
+def _storedstate_workaround(self, obj):
+    if isinstance(obj, (ops.framework.StoredDict, ops.framework.StoredList)):
+        return obj.to_json()
+    return _storedstate_workaround.default(obj)
+
+
+_storedstate_workaround.default = JSONEncoder().default
+JSONEncoder.default = _storedstate_workaround
+
```

Comment:
Looking around I found: https://www.geeksforgeeks.org/encoding-and-decoding-custom-objects-in-python-json/
Which describes what functions you can implement to pass in custom serialization to encode/decode to_json.
The problem is that we probably don't want all decoding to end up as StoredDict objects, and there wouldn't be another way to tell the difference between I want a dict and I want a StoredDict (in the data stream).

I'm reasonably happy with the `.to_json` methods, but I'm quite against monkey patching other libraries to get that behavior.

Reply:
I also don't like monkeypatching in general, and I tried to lump my responses together into one which addressed most of the comments, but the subclassed `JSONEncoder` is pretty ugly from a user perspective.

This patch is, essentially, "what's ugly for OF" versus "what's ugly for charm authors?"

One of the basics of duck typing is that it should behave the same. I hate that cpython does explicit type checking to determine whether it's serializable instead of looking for something in `__dict__` and that the alternative is a subclassed `JSONEncoder`, but we can't resolve that for now.

The gist of the patch is "if it calls itself a `Dict|Set|List`, it should behave like one". Requiring charm authors to use a custom `JSONEncoder` defeats the purpose. Additionally, it's not always clear when a charm is executing whether you've hit "pre-calculated" or unchanged data in `StoredState` or whether you generated new state which you are trying to pass to a relation.

We could strongly document that authors typecheck everything before blindly trying `json.dumps()` or that they should always ensure in code flow that it's one type or the other, but both of these lock users into an antipattern which is not Pythonic.

Reply:
I wonder why the type hint for `dumps` argument is `Any`:
```python
def dumps(obj: Any, ...)
```
If the type hint was correct then charm authors would have been much less confused with the existing StoredX as they are.

Reply:
I just noticed @benhoyt also mentioned the subclass route: https://github.com/canonical/operator/pull/572#issuecomment-885423135